clients/src/main/java/org/apache/kafka/common/MetricNameTemplate.java

```diff
@@ -26,27 +26,28 @@
  * A template for a MetricName. It contains a name, group, and description, as
  * well as all the tags that will be used to create the mBean name. Tag values
  * are omitted from the template, but are filled in at runtime with their
- * specified values.
+ * specified values. The order of the tags is maintained, if an ordered set
+ * is provided.
  */
 public class MetricNameTemplate {
     private final String name;
     private final String group;
     private final String description;
-    private Set<String> tags;
+    private LinkedHashSet<String> tags;
 
     public MetricNameTemplate(String name, String group, String description, Set<String> tags) {
```

Comment:
Someone could construct a HashSet that they added items to in the order "a" "b" "c", and pass it into MetricNameTemplate. It might then get manged here into the order "c" "b" "a". Should we have the MetricNameTemplate accept a LinkedHashSet(), instead of a Set?

Reply:
Doing this would also make the compile complain about all instances that instantiate MetricNameTemplates with a non-order-preserving Set, which I think is what we want. As it currently is implemented in the PR, I don't think the compile actually catches anything.

Reply:
I agree that enforcing a LinkedHashSet is harder to use. I'm a little worried about discoverability, though. As it is, there is no way to "discover" the "feature" that passing in an order-preserving map will, in fact, preserve order. It kind of feels obvious that it *should*, but it also kind of doesn't, since the constructor signature doesn't say anything about that.

Maybe add a comment to the javadocs for MetricNameTemplate() that if you pass in an order-preserving-Map, that it will preserve order? Or is that a useless comment, because it'll be easy to miss? As it is, the "ability to preserve order" is kind of an optional "power-user" feature, so maybe hiding the functionality in a comment is okay.

In your callers, maybe add a comment of "passing in a LinkedHashSet, so that order of tags is preserved in the docs"?

Reply:
Yeah, I think documenting it is better than making it much harder to use. Really, I think the only thing it affects is the generated documentation, and if somebody sees the order as wrong they (hopefully) will check the `MetricNameTemplate` JavaDoc.